cpp/tests/utilities/default_stream.cpp

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/utilities/default_stream.hpp>
+
+#include <cudf_test/default_stream.hpp>
+
+namespace cudf {
+namespace test {
+
+rmm::cuda_stream_view const get_default_stream() { return cudf::get_default_stream(); }
```

Comment:
Will this cause double definition of `cudf::test::get_default_stream()`? This is defined here, and also defined again in `identify_stream_usage.cpp`.

Reply:
It would if they were both compiled into the same library/executable, but the purpose of these is explicitly _not_ to be put into the same TU. The idea is that this file is compiled as part of libcudf's test utils, so it is the "official" definition of `cudf::test::get_default_stream`. When we want to ensure that all libcudf APIs are properly passing streams to everything they call (e.g. if I call `cudf::foo(..., stream)`, internally we want to make sure that `cudf::foo` calls `cudf::detail::foo(..., stream)` and doesn't forget to pass the stream through), we will run tests as `LD_PRELOAD=/path/to/identify_stream_usage TEST_EXECUTABLE`, which will result in the definition of `cudf::test::get_default_stream` in `identify_stream_usage` being used instead of the definition in `default_stream`. The definition in `identify_stream_usage` uses a custom stream defined as a function-local static, so we can then overload all CUDA APIs (like `cudaLaunchKernel`) to make sure that the stream that `cudaLaunchKernel` was called with is the same stream as `cudf::test::get_default_stream`. If it isn't, that implies that somewhere in libcudf we accidentally call a function without a stream (or we explicitly pass `cudf::get_default_stream`) instead of correctly forwarding the user-provided stream through.

Reply:
Humnn, this is the first time I know about `LD_PRELOAD`. I was confused in the first place as I didn't know it so I had no idea why we have repeated definitions. Now I understand it :)

However, this seems like a trick rather than an "official" technique when we overshadow a standard symbol by "preload" another version of that symbol contained in different libraries. But it seems the trick doesn't produce undefined behavior thus I'm fine with this.

Reply:
Yes, you could certainly think of it as a trick in some sense. I wouldn't advocate using it in production in 99% of cases. It turns out to be a perfect tool for what we want to validate with streams, though!